cql3/restrictions/statement_restrictions.cc

```diff
@@ -142,6 +142,50 @@ to_column_definition(const schema_ptr& schema, const ::shared_ptr<column_identif
 }
 #endif
 
+/// Every token, or if no tokens, an EQ/IN of every single PK column.
+static std::vector<expr::expression> extract_partition_range(
+        const expr::expression& where_clause, schema_ptr schema) {
+    using namespace expr;
+    struct {
+        std::optional<expression> tokens;
+        std::unordered_map<const column_definition*, expression> single_column;
+
```

Comment:
We are now entering an area where every single query will hit the changes. So we have to be sensitive about performance. Please run `perf_simple_query --smp 1` before and after. Note it also reports allocations per queries.

In particular, I expect this map to add several allocations. I also think it can be moved to the prepared statement, since it doesn't involve query_options.

Reply:
(I jumped into conclusions here, probably it already is part of the prepared statement; but running the benchmark is good regardless)

Reply:
Correct, this is at prepare time.  I ran `perf_simple_query --smp 1` before and after:

Before (c1daf2bb2):
```
36686.55 tps
36915.25 tps
36784.31 tps
36803.77 tps
36734.69 tps

median 36784.31
median absolute deviation: 49.62
maximum: 36915.25
minimum: 36686.55
```
After (c48ffa1):
```
29409.18 tps
29358.62 tps
29381.55 tps
29457.00 tps
29444.45 tps

median 29409.18
median absolute deviation: 35.28
maximum: 29457.00
minimum: 29358.62
```

Reply:
I didn't see any references to allocation in the output. :shrug:

Reply:
Rebase on master and you will.

That's a huge regression, so something got broken.

Also, what are you running this on, a 6502? My machine gets 185k ops/sec.

Reply:
> Also, what are you running this on, a 6502?

That was dev mode, but release is also slow: only 70Ktps after the regression.

> My machine gets 185k ops/sec.

I bet your machine has actually run scylla_setup successfully once.

Reply:
I've never run scylla_setup on my machine. perf_simple_query --smp 1 only depends on the cpu core.

Run performance tests in release mode, dev mode is worthless for those.

Reply:
After replacing a std::transform with a manual for-loop, the latest result is 91378.47 tps (100 allocs/op, 20 tasks/op).  Even better than master! :muscle: I guess I was mistaken about the extra 10 allocs/op slowing us down.